pkg/security/ebpf/c/dentry_resolver.h

```diff
@@ -0,0 +1,333 @@
+        map_value.name[0] = 0;
+        map_value.parent.mount_id = 0;
+        map_value.parent.ino = 0;
+        bpf_map_update_elem(&pathnames, &next_key, &map_value, BPF_ANY);
+    }
+
+    // prepare for the next iteration
+    input->dentry = d_parent;
+    input->key = next_key;
+    return DR_MAX_ITERATION_DEPTH;
+}
+
+SEC("kprobe/dentry_resolver_kern")
+int kprobe__dentry_resolver_kern(struct pt_regs *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_ANY);
+    if (!syscall)
+        return 0;
+
+    syscall->resolver.iteration++;
+    syscall->resolver.ret = resolve_dentry_tail_call(&syscall->resolver);
+
+    if (syscall->resolver.ret > 0) {
+        if (syscall->resolver.iteration < DR_MAX_TAIL_CALL && syscall->resolver.key.ino != 0) {
+            bpf_tail_call(ctx, &dentry_resolver_kprobe_progs, DR_KPROBE);
+        }
+
+        syscall->resolver.ret += DR_MAX_ITERATION_DEPTH * (syscall->resolver.iteration - 1);
+    }
+
+    if (syscall->resolver.callback >= 0) {
+        bpf_tail_call(ctx, &dentry_resolver_kprobe_callbacks, syscall->resolver.callback);
+    }
+    return 0;
+}
+
+SEC("tracepoint/dentry_resolver_kern")
+int tracepoint__dentry_resolver_kern(void *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_ANY);
+    if (!syscall)
+        return 0;
+
+    syscall->resolver.iteration++;
+    syscall->resolver.ret = resolve_dentry_tail_call(&syscall->resolver);
+
+    if (syscall->resolver.ret > 0) {
+        if (syscall->resolver.iteration < DR_MAX_TAIL_CALL && syscall->resolver.key.ino != 0) {
+            bpf_tail_call(ctx, &dentry_resolver_tracepoint_progs, DR_TRACEPOINT);
+        }
+
+        syscall->resolver.ret += DR_MAX_ITERATION_DEPTH * (syscall->resolver.iteration - 1);
+    }
+
+    if (syscall->resolver.callback >= 0) {
+        bpf_tail_call(ctx, &dentry_resolver_tracepoint_callbacks, syscall->resolver.callback);
+    }
+    return 0;
+}
+
+struct dr_erpc_state_t {
+    char *userspace_buffer;
+    struct path_key_t key;
+    int ret;
+    int iteration;
+    u32 buffer_size;
+    u16 cursor;
+};
+
+struct bpf_map_def SEC("maps/dr_erpc_state") dr_erpc_state = {
+    .type = BPF_MAP_TYPE_ARRAY,
+    .key_size = sizeof(u32),
+    .value_size = sizeof(struct dr_erpc_state_t),
+    .max_entries = 1,
+    .pinning = 0,
+    .namespace = "",
+};
+
+SEC("kprobe/dentry_resolver_erpc")
+int kprobe__dentry_resolver_erpc(struct pt_regs *ctx) {
+    u32 key = 0;
+    struct path_leaf_t *map_value = 0;
+
+    struct dr_erpc_state_t *state = bpf_map_lookup_elem(&dr_erpc_state, &key);
+    if (state == NULL) {
+        return 0;
+    }
+
+    state->iteration++;
+
+#pragma unroll
+    for (int i = 0; i < DR_MAX_ITERATION_DEPTH; i++)
+    {
+        map_value = bpf_map_lookup_elem(&pathnames, &state->key);
+        if (map_value == NULL)
+            goto exit;
+
+        // make sure we do not write outside of the provided buffer
+        if (state->cursor + 16 >= state->buffer_size)
```

Comment:
s/16/sizeof(state->key)/ ? does it make sense ?

Reply:
👍🏻